Separates edgecolor from hatchcolor (#28104)
matplotlib/matplotlib: doc/users/next_whats_new/separated_hatchcolor.rst

    ax.add_patch(patch1)

    # When hatchcolor is not specified, it matches edgecolor
    # In this case, hatchcolor is green
    patch2 = Rectangle((0.6, 0.1), 0.3, 0.3, edgecolor='green', linewidth=2,
                       hatch='//', facecolor='none')
    ax.add_patch(patch2)

    # If both hatchcolor and edgecolor are not specified
    # it will default to the 'patch.edgecolor' rcParam, which is black by default
    # In this case, hatchcolor is black
    patch3 = Rectangle((0.1, 0.6), 0.3, 0.3, hatch='//')
    ax.add_patch(patch3)

    # When using `hatch.color` in the `rcParams`
    # edgecolor will now not overwrite hatchcolor
    # In this case, hatchcolor is black
    with plt.rc_context({'hatch.color': 'black'}):
        patch4 = Rectangle((0.6, 0.6), 0.3, 0.3, edgecolor='blue', linewidth=2,
                           hatch='//', facecolor='none')

    # hatchcolor is black (it uses the `hatch.color` rcParam value)
    assert patch4.get_hatchcolor() == mpl.colors.to_rgba('black')
    patch4.set_edgecolor('blue')
    # hatchcolor is still black (here, it does not update when edgecolor changes)
    assert patch4.get_hatchcolor() == mpl.colors.to_rgba('black')

>     ax.add_patch(patch4)

Author: The asserts strengthen the comments, and the comments themselves are not sufficient I feel. If it feels unnecessary here too, then it can be removed

Separated ``hatchcolor`` from ``edgecolor``
-------------------------------------------

`~matplotlib.patches.Patch` gained a new *hatchcolor* parameter to explicitly
control hatch colors. Previously, hatch colors were the same as edge colors,
with a fallback to :rc:`hatch.color` if the patch did not have an edge color.

Fallback Logic
~~~~~~~~~~~~~~
When the *hatchcolor* parameter is specified, it will be used for the hatch.

> If it is not specified, it will fallback to using :rc:`hatch.color`.

Reviewer: ```suggestion
If it is not specified, it will fall back to using :rc:`hatch.color`.
```

> The special value 'edge' takes over the patch edgecolor, with a fallback to

Reviewer: ```suggestion
The special value 'edge' uses the patch edgecolor, with a fall back to
```

Reviewer: Oops, my mistake; the `fallback` here is a noun, not a verb, and shouldn't get the space added.

:rc:`patch.edgecolor` if the patch edgecolor is 'none'.

.. plot::
    :include-source: true
    :alt: Example of using the hatchcolor parameter in a Rectangle

    import matplotlib as mpl
    import matplotlib.pyplot as plt
    from matplotlib.patches import Rectangle

    fig, ax = plt.subplots()

    # hatchcolor can now be controlled using the `hatchcolor` parameter
    # In this case, hatchcolor is orange
    patch1 = Rectangle((0.1, 0.1), 0.3, 0.3, edgecolor='red', linewidth=2,
                      hatch='//', hatchcolor='orange')
    ax.add_patch(patch1)

    # When hatchcolor is not specified, it matches edgecolor
    # In this case, hatchcolor is green
    patch2 = Rectangle((0.6, 0.1), 0.3, 0.3, edgecolor='green', linewidth=2,
                       hatch='//', facecolor='none')
    assert patch2._hatch_color == 'edge'

>     assert patch2.get_hatchcolor() == mpl.colors.to_rgba('green')

Reviewer: Do we really need asserts for the example?

Author: Yup, asserts seem unnecessary

Reviewer 2: I've mixed feelings: Typically not. Here it's a way to demonstrate that the default hatchcolor is "edge" and that the resulting color will thus be "green". But the comment above should be sufficient.

Author: Yes, the comments should be sufficient

Separated ``hatchcolor`` from ``edgecolor``
-------------------------------------------

`~matplotlib.patches.Patch` gained a new *hatchcolor* parameter to explicitly
control hatch colors. Previously, hatch colors were the same as edge colors,
with a fallback to :rc:`hatch.color` if the patch did not have an edge color.

Fallback Logic
~~~~~~~~~~~~~~
When the *hatchcolor* parameter is specified, it will be used for the hatch.
If it is not specified, it will fallback to using :rc:`hatch.color`.

The special value 'edge' takes over the patch edgecolor, with a fallback to
:rc:`patch.edgecolor` if the patch edgecolor is 'none'.

If the patch hatchcolor is 'edge' then hatchcolor matches edgecolor, that is, it will
be updated if edgecolor is changed (for example: by calling *set_edgecolor()*).
But if hatchcolor is explicitly set (for example: by calling *set_hatchcolor()*

> or by using *hatch.color* rcParam), it will not be updated when edgecolor changes.

Reviewer 2: I find this confusing. Logically, `hatchcolor='edge'` just always uses the the edgecolor, but that's already expressed in the previous paragraph. I don't know what information is needed on top of that. Do we need this paragraph at all? It reads a bit like an implementation detail.

Reviewer 2: I've unresolved this, because I don't see any action relating the paragraph or a comment why it should stay as is.

Author: I'm sorry, missed this 🤦‍♂️

> Separated ``hatchcolor`` from ``edgecolor``
> -------------------------------------------
> 
> The ``hatchcolor`` parameter for `~matplotlib.patches.Patch` objects has been
> separated from the ``edgecolor`` parameter. This allows for using different colors
> for hatches and edges by explicitly setting the ``hatchcolor`` parameter.

Reviewer 2: ```suggestion
Separated ``hatchcolor`` from ``edgecolor``
-------------------------------------------

`~matplotlib.patches.Patch` gained a new *hatchcolor* parameter to explicitly
control hatch colors. Previously, hatch colors were the same as edge colors,
with a fallback to :rc:`hatch.color` if the patch did not have an edge color.
```

Fallback Logic
~~~~~~~~~~~~~~
When the *hatchcolor* parameter is specified, it will be used for the hatch.
If it is not specified, it will fall back to using :rc:`hatch.color`.

The special value 'edge' uses the patch edgecolor, with a fallback to

> :rc:`patch.edgecolor` if the patch edgecolor is 'none'.

Reviewer 2: Um, no? The value can be used via rcParam or kwarg, so narrowing down is not helpful. We could leave out the subsection "Fallback logic". The paragraphs work well with out that title. Also the subsequent plot is general and not tied to the fallback logic.

> If this is a valid color, it will be used for the hatch.
> 
> If the *hatch.color* rcParam is not set, then its default value will be used,
> which is *inherit*. In this case, *hatchcolor* will try to inherit from *edgecolor*
> of the patch if it is specified. If *edgecolor* is also not specified,
> *hatchcolor* will fallback to :rc:`patch.edgecolor`.

Reviewer 2: ```suggestion

The special value 'inherit' takes over the patch edgecolor, with a fallback to
:rc:`patch.edgecolor` if the patch edgecolor is 'none'.
```
Is this how it's implemented? - The logic/priority is: Synchronize hatchcolor with edgecolor, but if the patch does not have a visible edge, perfer to fall through to `rcParams['patch.edgecolor']` - we usually don't want the hatches to vanish.
